src/plugins/chatMarkers/ChatMarkersPlugin.ts

```diff
@@ -0,0 +1,333 @@
+         this.pluginAPI.Log.debug('contact has bare jabber id. Yaay! =)');
+         return Promise.resolve(true);
+      } else {
+         // if full JID, sender SHOULD try to determine if recipient supports chat markers
+         this.pluginAPI.Log.debug('contact has full jabber id.');
+         return this.pluginAPI.getDiscoInfoRepository().hasFeature(jid, [Namespace.get(CHATMARKERS)]).then(() => {
+            this.pluginAPI.Log.debug('contact supports chat markers. Yaay! =)');
+            return Promise.resolve(true);
+         }).catch(() => {
+            this.pluginAPI.Log.debug('contact does not support chat markers. =(');
+            return Promise.resolve(false);
+         });
+      }
+   }
+
+   // add "markable" element according to XEP-0333
+   private addMarkable(xmlStanza: Strophe.Builder, msg: Message) {
+      xmlStanza.c(MARKABLE, {
+         xmlns: Namespace.get(CHATMARKERS)
+      }).up();
+   }
+
+   // send "received" message according to XEP-0333
+   private sendReceived(lastReceivedMsgId: string) {
+      this.pluginAPI.send($msg({
+         to: this.data.get('lastReceivedFrom')
+      }).c(RECEIVED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastReceivedMsgId
+      }));
+   }
+
+   // send "displayed" message according to XEP-0333
+   private sendDisplayed(lastDisplayedMsgId: string) {
+      this.pluginAPI.send($msg({
+         to: this.data.get('lastDisplayedFrom')
+      }).c(DISPLAYED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastDisplayedMsgId
+      }));
+   }
+
+   // send "acknowledged" message according to XEP-0333
+   private sendAcknowledged(lastAcknowledgedMsgId: string) {
+      this.pluginAPI.send($msg({
+         to: this.data.get('lastAcknowledgedFrom')
+      }).c(ACKNOWLEDGED, {
+         xmlns: Namespace.get(CHATMARKERS),
+         id: lastAcknowledgedMsgId
+      }));
+   }
+
+   // set "chatMarkersAcknowledged" on all messages
+   private setMessagesChatMarkersAcknowledged(lastAcknowledgedMsgUid: string) {
+      let transcript = this.pluginAPI.getContact(new JID(this.data.get('lastAcknowledgedFrom'))).getChatWindow().getTranscript();
+      let msg = transcript.getMessage(lastAcknowledgedMsgUid) as Message;
+      while (!!msg && !msg.isChatMarkersAcknowledged()) {
+         if (msg.getDirection() === DIRECTION.IN) {
+            msg.chatMarkersAcknowledged();
+         }
+         try {
+            msg = transcript.getMessage(msg.getNextId()) as Message;
+         } catch (error) {
+            break;
+         }
+      }
+   }
+
+   // add "markable" to message according to XEP-0333
+   private preSendMessageStanzaProcessor = (msg: Message, xmlStanza: Strophe.Builder): Promise<any> => {
+      this.supportsChatMarkers(msg.getPeer()).then(() => {
+         if (msg.getType() !== Message.MSGTYPE.GROUPCHAT) {
+            this.addMarkable(xmlStanza, msg);
+         }
+      });
+      return Promise.resolve([msg, xmlStanza]);
```

Comment:
Modifications from addMarkable are lost at this point. Return this.supportsChatMarkers and return the array inside the then function.